### ISSUE TITLE
S3 client getObject() throws exception on 304 ("Not Modified")

### ISSUE THREAD

**Issue by anelsonotto** [aws/aws-sdk-java-v2#428]
When getting an object from an S3 bucket with "ifNoneMatch", an exception is thrown if the eTag matches.

ResponseInputStream<GetObjectResponse> response =  s3Client.getObject(getObjectRequest, StreamingResponseHandler.toInputStream());

## Expected Behavior
Not throwing an exception, instead only return the status code 304 and null InputStream when eTag matches.

See https://forums.aws.amazon.com/thread.jspa?threadID=77995&tstart=0

## Your Environment
* AWS Java SDK version used: 2.0.0-preview-8

**Comment by jrduncans** [comment]
I agree that an exception is pretty heavy-weight for this situation, which is unfortunate. It's also not made clear at all in the documentation, and it's not a specific exception class, so you have to catch all S3Exception and then check the status code to be able to use it. It's a very disappointing step backwards from the V1 SDK (even if returning null is a bit ugly, it's at least cheap and easy to document).

**Comment by mrfilipenko** [comment]
Would also like this to be fixed.
